# Creating a Sentiment Analysis Web App
## Using PyTorch and SageMaker

_Deep Learning Nanodegree Program | Deployment_

---

Now that we have a basic understanding of how SageMaker works we will try to use it to construct a complete project from end to end. Our goal will be to have a simple web page which a user can use to enter a movie review. The web page will then send the review off to our deployed model which will predict the sentiment of the entered review.

## Instructions

Some template code has already been provided for you, and you will need to implement additional functionality to successfully complete this notebook. You will not need to modify the included code beyond what is requested. Sections that begin with '**TODO**' in the header indicate that you need to complete or implement some portion within them. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `# TODO: ...` comment. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions for you to answer which relate to the task and your implementation. Each section where you will answer a question is preceded by a '**Question:**' header. Carefully read each question and provide your answer below the '**Answer:**' header by editing the Markdown cell.

> **Note**: Code and Markdown cells can be executed using the **Shift+Enter** keyboard shortcut. In addition, a cell can be edited by typically clicking it (double-click for Markdown cells) or by pressing **Enter** while it is highlighted.

## General Outline

Recall the general outline for SageMaker projects using a notebook instance.

1. Download or otherwise retrieve the data.
2. Process / Prepare the data.
3. Upload the processed data to S3.
4. Train a chosen model.
5. Test the trained model (typically using a batch transform job).
6. Deploy the trained model.
7. Use the deployed model.

For this project, you will be following the steps in the general outline with some modifications. 

First, you will not be testing the model in its own step. You will still be testing the model, however, you will do it by deploying your model and then using the deployed model by sending the test data to it. One of the reasons for doing this is so that you can make sure that your deployed model is working correctly before moving forward.

In addition, you will deploy and use your trained model a second time. In the second iteration you will customize the way that your trained model is deployed by including some of your own code. In addition, your newly deployed model will be used in the sentiment analysis web app.

#### Student Comment: 
Throughout this notebook there are a number of comments labeled  **Take-A-Look**. 
These comments explore the data and object types and how they are being transformed.  

## Step 1: Downloading the data

As in the XGBoost in SageMaker notebook, we will be using the [IMDb dataset](http://ai.stanford.edu/~amaas/data/sentiment/)

> Maas, Andrew L., et al. [Learning Word Vectors for Sentiment Analysis](http://ai.stanford.edu/~amaas/data/sentiment/). In _Proceedings of the 49th Annual Meeting of the Association for Computational Linguistics: Human Language Technologies_. Association for Computational Linguistics, 2011.

In [1]:
# Notebook version: 12-17-20 AWS
# Invoke URL: https://lfmrtiwpfh.execute-api.us-east-2.amazonaws.com/production
#https://github.com/udacity/sagemaker-deployment/blob/master/Project/SageMaker%20Project.ipynb

In [2]:
#%mkdir ../data

#!wget -O ../data/aclImdb_v1.tar.gz http://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
#!tar -zxf ../data/aclImdb_v1.tar.gz -C ../data

In [3]:
!pwd

/home/ec2-user/SageMaker/sagemaker-deployment/Project


In [4]:
#!pip install sagemaker==1.72.0

## Step 2: Preparing and Processing the data

Also, as in the XGBoost notebook, we will be doing some initial data processing. The first few steps are the same as in the XGBoost example. To begin with, we will read in each of the reviews and combine them into a single input structure. Then, we will split the dataset into a training set and a testing set.

In [5]:
import os
import glob
from timeit import default_timer as timer
from collections import Counter, OrderedDict
from sagemaker.pytorch import PyTorch
import collections
import itertools 
#import sagemaker_containers
import helpers
from helpers import oftype
from sklearn.metrics import accuracy_score
#import transformers
import pprint
import random
import sagemaker

def read_imdb_data(data_dir='../data/aclImdb'):
    data = {}
    labels = {}
    
    for data_type in ['train', 'test']:
        data[data_type] = {}
        labels[data_type] = {}
        
        for sentiment in ['pos', 'neg']:
            data[data_type][sentiment] = []
            labels[data_type][sentiment] = []
            
            path = os.path.join(data_dir, data_type, sentiment, '*.txt')
            files = glob.glob(path)
            
            for f in files:
                with open(f) as review:
                    data[data_type][sentiment].append(review.read())
                    # Here we represent a positive review by '1' and a negative review by '0'
                    labels[data_type][sentiment].append(1 if sentiment == 'pos' else 0)
                    
            assert len(data[data_type][sentiment]) == len(labels[data_type][sentiment]), \
                    "{}/{} data size does not match labels size".format(data_type, sentiment)
                
    return data, labels

In [6]:
!python --version
!pip --version
print (sagemaker.__version__)

Python 3.6.10 :: Anaconda, Inc.
pip 20.0.2 from /home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/pip (python 3.6)
1.72.0


In [7]:
data, labels = read_imdb_data()
print("IMDB reviews: train = {} pos / {} neg, test = {} pos / {} neg".format(
            len(data['train']['pos']), len(data['train']['neg']),
            len(data['test']['pos']), len(data['test']['neg'])))

IMDB reviews: train = 12500 pos / 12500 neg, test = 12500 pos / 12500 neg


> **Take-A-Look:** The next two code blocks explore the two dictionaries created above called 'data' and 'labels'. The 'data' dictionary has two levels of nesting. The first layer has 2 keys: 'train' and 'test'. The values are the second level of the dictionary, with the values being key-value pairs themselves. The second level keys are 'neg' and 'pos' with their corresponding values the reviews. This is illustrated below:

In [8]:
#show dictionary structure: 
pprint.pprint(data,depth =2,indent = 4)
#Note that data is a nested dictionary: 

{'test': {'neg': [...], 'pos': [...]}, 'train': {'neg': [...], 'pos': [...]}}


In [9]:
#Take a look at a more detailed look at the dictionary structure:  
def dictList(data,N):
    #d = collections.OrderedDict(data)
    d = (data)
    x = itertools.islice(d.items(),0,N) 

    for key, value in x: 
        display ( str(key) )
        for key1,value1 in value.items(): 
            print ('\t' + str(key1) + ' : ' + str(str(value1)[0:30]))
dictList(data,2)

'train'

	pos : ['So what constitutes a real i
	neg : ['This movie was crap with a c


'test'

	pos : ["Giant Robot was the most pop
	neg : ["Not much actually happens in


Now that we've read the raw training and testing data from the downloaded dataset, we will combine the positive and negative reviews and shuffle the resulting records.

In [10]:
from sklearn.utils import shuffle

def prepare_imdb_data(data, labels):
    """Prepare training and test sets from IMDb movie reviews."""
    
    #Combine positive and negative reviews and labels
    data_train = data['train']['pos'] + data['train']['neg']
    data_test = data['test']['pos'] + data['test']['neg']
    labels_train = labels['train']['pos'] + labels['train']['neg']
    labels_test = labels['test']['pos'] + labels['test']['neg']
    #Shuffle reviews and corresponding labels within training and test sets
    data_train, labels_train = shuffle(data_train, labels_train)
    data_test, labels_test = shuffle(data_test, labels_test)
    # Return a unified training data, test data, training labels, test labets
    return data_train, data_test, labels_train, labels_test

In [11]:
train_X, test_X, train_y, test_y = prepare_imdb_data(data, labels)
print("IMDb reviews (combined): train = {}, test = {}".format(len(train_X), len(test_X)))

IMDb reviews (combined): train = 25000, test = 25000


> **Take-A-Look:** Note that the input to the prepare_imdb_data method above are two dictionaries while the returned output are 4 lists: train_X, test_X,train_y,test_y.

In [12]:
oftype(train_X),oftype(test_X)
oftype(train_y)
oftype(test_y)

List of length:  25000
List of length:  25000
List of length:  25000
List of length:  25000


Now that we have our training and testing sets unified and prepared, we should do a quick check and see an example of the data our model will be trained on. This is generally a good idea as it allows you to see how each of the further processing steps affects the reviews and it also ensures that the data has been loaded correctly.

The first step in processing the reviews is to make sure that any html tags that appear should be removed. In addition we wish to tokenize our input, that way words such as *entertained* and *entertaining* are considered the same with regard to sentiment analysis.

In [13]:
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import *

import re
from bs4 import BeautifulSoup

def review_to_words(review):
    nltk.download("stopwords", quiet=True)
    stemmer = PorterStemmer()
    
    text = BeautifulSoup(review, "html.parser").get_text() # Remove HTML tags
    text = re.sub(r"[^a-zA-Z0-9]", " ", text.lower()) # Convert to lower case
    words = text.split() # Split string into words
    words = [w for w in words if w not in stopwords.words("english")] # Remove stopwords
    words = [PorterStemmer().stem(w) for w in words] # stem
    
    return words

> **Take-A-Look:** Below is an example of how the method 'review_to_words' transforms the input. Note that the input is a string and the output is a list. 

In [14]:
input_text = '\this is a - entertaining / - reviewing at a loss'
oftype(input_text)
output_text = review_to_words(input_text)
print (output_text)
oftype(output_text)

string of length:  48 , 	his is a - entertaining / - reviewing at a loss
['entertain', 'review', 'loss']
List of length:  3


The `review_to_words` method defined above uses `BeautifulSoup` to remove any html tags that appear and uses the `nltk` package to tokenize the reviews. As a check to ensure we know how everything is working, try applying `review_to_words` to one of the reviews in the training set.

**Question:** Above we mentioned that `review_to_words` method removes html formatting and allows us to tokenize the words found in a review, for example, converting *entertained* and *entertaining* into *entertain* so that they are treated as though they are the same word. What else, if anything, does this method do to the input?

**Answer:**
review_to_words does the following: removes HTML tags, converts to lower case, splits the string into words, removes stopwords and stems the words. 

The method below applies the `review_to_words` method to each of the reviews in the training and testing datasets. In addition it caches the results. This is because performing this processing step can take a long time. This way if you are unable to complete the notebook in the current session, you can come back without needing to process the data a second time.

In [15]:
import pickle

cache_dir = os.path.join("../cache", "sentiment_analysis")  # where to store cache files
os.makedirs(cache_dir, exist_ok=True)  # ensure cache directory exists

def preprocess_data(data_train, data_test, labels_train, labels_test,
                    cache_dir=cache_dir, cache_file="preprocessed_data.pkl"):
    """Convert each review to words; read from cache if available."""

    # If cache_file is not None, try to read from it first
    cache_data = None
    if cache_file is not None:
        try:
            with open(os.path.join(cache_dir, cache_file), "rb") as f:
                cache_data = pickle.load(f)
            print("Read preprocessed data from cache file:", cache_file)
        except:
            pass  # unable to read from cache, but that's okay
    
    # If cache is missing, then do the heavy lifting
    if cache_data is None:
        # Preprocess training and test data to obtain words for each review
        #words_train = list(map(review_to_words, data_train))
        #words_test = list(map(review_to_words, data_test))
        words_train = [review_to_words(review) for review in data_train]
        words_test = [review_to_words(review) for review in data_test]
        
        # Write to cache file for future runs
        if cache_file is not None:
            cache_data = dict(words_train=words_train, words_test=words_test,
                              labels_train=labels_train, labels_test=labels_test)
            with open(os.path.join(cache_dir, cache_file), "wb") as f:
                pickle.dump(cache_data, f)
            print("Wrote preprocessed data to cache file:", cache_file)
    else:
        # Unpack data loaded from cache file
        words_train, words_test, labels_train, labels_test = (cache_data['words_train'],
                cache_data['words_test'], cache_data['labels_train'], cache_data['labels_test'])
    
    return words_train, words_test, labels_train, labels_test

Preprocess the data. This applies review_to_words to each of the reviews, stripping and stemming characters as needed. 

> **Take-A-Look:** The preprocess_data method leaves the data types unchanged. List input and List output. 

In [16]:
oftype(train_X)

List of length:  25000


In [17]:
# Preprocess data
train_X, test_X, train_y, test_y = preprocess_data(train_X, test_X, train_y, test_y)

Read preprocessed data from cache file: preprocessed_data.pkl


In [18]:
oftype(train_X)

List of length:  25000


In [19]:
len(train_X[0])

402

In [20]:
# Option to cut down on dataset size for faster execution during debug: 

def mysample(df,sample_size):
    print (oftype(df))
    df_s=random.sample(list(df), sample_size)
    print (oftype(df_s))
    return df_s

In [21]:
#setting up randome sample for quicker debug 
#train_X = mysample(train_X,500)
#train_y = mysample(train_y,500)


In [22]:
#test_X = mysample(test_X,500)
#test_y = mysample(test_y,500)

## Transform the data

In the XGBoost notebook we transformed the data from its word representation to a bag-of-words feature representation. For the model we are going to construct in this notebook we will construct a feature representation which is very similar. To start, we will represent each word as an integer. Of course, some of the words that appear in the reviews occur very infrequently and so likely don't contain much information for the purposes of sentiment analysis. The way we will deal with this problem is that we will fix the size of our working vocabulary and we will only include the words that appear most frequently. We will then combine all of the infrequent words into a single category and, in our case, we will label it as `1`.

Since we will be using a recurrent neural network, it will be convenient if the length of each review is the same. To do this, we will fix a size for our reviews and then pad short reviews with the category 'no word' (which we will label `0`) and truncate long reviews.

### (TODO) Create a word dictionary

To begin with, we need to construct a way to map words that appear in the reviews to integers. Here we fix the size of our vocabulary (including the 'no word' and 'infrequent' categories) to be `5000` but you may wish to change this to see how it affects the model.

> **TODO:** Complete the implementation for the `build_dict()` method below. Note that even though the vocab_size is set to `5000`, we only want to construct a mapping for the most frequently appearing `4998` words. This is because we want to reserve the special labels `0` for 'no word' and `1` for 'infrequent word'.

In [23]:
import numpy as np

def build_dict(data, vocab_size = 5000):
    """Construct and return a dictionary mapping each of the most frequently appearing words to a unique integer."""
    
    # TODO: Determine how often each word appears in `data`. Note that `data` is a list of sentences and that a
    #       sentence is a list of words.
    word_count = {} # A dict storing the words that appear in the reviews along with how often they occur
        
    for review in data:
         for word in review:
           
                word_count[word] = word_count[word] +1 if word in word_count else 1 
                #print ("word: ",word, "count: ",word_count[word])
    x=0
    print ("The output of build_dict is a dictionary of word,sorted importance:\n")
    for key,value in word_count.items(): 
        x=x+1 
        print ("{}: {}".format(key,value))
        if x> 5:
            break
    # TODO: Sort the words found in `data` so that sorted_words[0] is the most frequently appearing word and
    #       sorted_words[-1] is the least frequently appearing word.
    
    word_count_sorted = sorted(word_count.items(), key=(lambda item: item[1]), reverse=True)
  
    sorted_words = [item[0] for item in word_count_sorted] 
    word_dict = {} # This is what we are building, a dictionary that translates words into integers
    for idx, word in enumerate(sorted_words[:vocab_size - 2]): # The -2 is so that we save room for the 'no word'
           word_dict[word] = idx + 2                              # 'infrequent' labels

    return word_dict

> **Take-A-Look:** Let's take a look at word_dict. It is a dictionary of length 4998 + 2. The dictionary key is a word while the value for that key is its sorted position in terms of word count. 

In [24]:
word_dict=build_dict(train_X)

The output of build_dict is a dictionary of word,sorted importance:

dug: 28
watch: 13940
tonight: 96
honestli: 453
think: 8913
must: 3250


In [25]:
oftype(word_dict)

Dictionary of length:  4998


In [26]:
# TODO: Use this space to determine the five most frequently appearing words in the training set.
print (list(word_dict.items())[:5])

[('movi', 2), ('film', 3), ('one', 4), ('like', 5), ('time', 6)]


> **Take-A-Look:** Let's confirm that 'movi' is '2' in the word_dict: 

In [27]:
key_list = list(word_dict.keys())
value_list=list(word_dict.values())
position = value_list.index(2)
print (key_list[position]) 

movi


**Answer:** All of the words are common words and its possible they are the top 5 most used. 

### Save `word_dict`

Later on when we construct an endpoint which processes a submitted review we will need to make use of the `word_dict` which we have created. As such, we will save it to a file now for future use.

In [28]:
data_dir = '../data/pytorch' # The folder we will use for storing data
if not os.path.exists(data_dir): # Make sure that the folder exists
    os.makedirs(data_dir)

In [29]:
with open(os.path.join(data_dir, 'word_dict.pkl'), "wb") as f:
    pickle.dump(word_dict, f)

### Transform the reviews

Now that we have our word dictionary which allows us to transform the words appearing in the reviews into integers, it is time to make use of it and convert our reviews to their integer sequence representation, making sure to pad or truncate to a fixed length, which in our case is `500`.

In [30]:
def convert_and_pad(word_dict, sentence, pad=500):
    NOWORD = 0 # We will use 0 to represent the 'no word' category
    INFREQ = 1 # and we use 1 to represent the infrequent words, i.e., words not appearing in word_dict
    
    working_sentence = [NOWORD] * pad
    
    for word_index, word in enumerate(sentence[:pad]):
        if word in word_dict:
            working_sentence[word_index] = word_dict[word]
        else:
            working_sentence[word_index] = INFREQ
            
    return working_sentence, min(len(sentence), pad)

def convert_and_pad_data(word_dict, data, pad=500):
    result = []
    lengths = []
    
    for sentence in data:
        converted, leng = convert_and_pad(word_dict, sentence, pad)
        result.append(converted)
        lengths.append(leng)
        
    return np.array(result), np.array(lengths)

 > **Take-A-Look:** Note that train_X is being converted from a LIST of 25,0000 reviews, with each review a single string,  to a numpy ARRAY of 25,000x500. In other words, now every 1 review has had 500 of its words translated into 500 integers. The integers were assigned from the word_dict. The value of the integer is determine by the sorted frequency of its usage. 

In [31]:
print ("Prior to conversion: ".format(type(train_X)))
oftype(train_X)
train_X_oldtype = type(train_X)
train_X, train_X_len = convert_and_pad_data(word_dict, train_X)
print ("After List conversion to Numpy Array: ".format(type(train_X)))
oftype(train_X)

Prior to conversion: 
List of length:  25000
After List conversion to Numpy Array: 
numpy (25000, 500)


In [32]:
# Now for test dataset. 
test_X, test_X_len = convert_and_pad_data(word_dict, test_X)

In [33]:
# Use this cell to examine one of the processed reviews to make sure everything is working as intended.
display (train_X[:1])

array([[   1,   12, 3706, 1165,   30,  130,  776,   44,  158,  136,    6,
         135,  254,  531,  848,    3,  254,   24,  304,    1,  783,   21,
         790,    1,    1, 1304,  160,   78,  948,   65,  307, 1187,  125,
          73,   67,  281, 2481,  141,  288,  463,    9,   13,  160,  714,
         356,  403,   67,  629, 1453,   48,  531,  748, 3991,  299,   85,
         271,   24,  251,   24,   32, 2295,    5,  430,  282,   51,   65,
          31,  412,  266,    3,  247, 2580, 2554,  948,   85,  461,  613,
         264,  841,  171,  448,  154,  674,   28, 1024,  776,  101,    3,
         264,   16,  264, 2697,   28,  777,  448,   12, 3706,  748,  108,
         139,   93,  588,  643,  255, 2811,   20,   65,    1,   12, 1406,
         811,  249, 1740, 4248, 1974,  240,  403,    1,    1, 1942,   35,
         245,  152, 1927,  674,  436,  157,  255,  207,  219, 2811,    1,
        1762,   82,  448,  303, 2004, 1845,  461,  161,  213,  154,  171,
        1094,   68,  878,  467,  160, 

>**Take-A-Look** In summary, train_y is a list of 25000, train_X_len is a numpy array of 25000 and train_X is a numpy array of 25000 x 500. These will be concatenated and saved to a *.csv file below. This file will then be uploaded to S3. 

In [34]:
oftype(train_y)
oftype(train_X_len)
oftype(train_X)

List of length:  25000
numpy (25000,)
numpy (25000, 500)


As a quick check to make sure that things are working as intended, check to see what one of the reviews in the training set looks like after having been processeed. Does this look reasonable? What is the length of a review in the training set?

**Question:** In the cells above we use the `preprocess_data` and `convert_and_pad_data` methods to process both the training and testing set. Why or why not might this be a problem?

**Answer:** The truncation and padding could introduce some errors. Truncation has looss of data. 

## Step 3: Upload the data to S3

As in the XGBoost notebook, we will need to upload the training dataset to S3 in order for our training code to access it. For now we will save it locally and we will upload to S3 later on.

### Save the processed training dataset locally

It is important to note the format of the data that we are saving as we will need to know it when we write the training code. In our case, each row of the dataset has the form `label`, `length`, `review[500]` where `review[500]` is a sequence of `500` integers representing the words in the review.

In [35]:
import pandas as pd
    
pd.concat([pd.DataFrame(train_y), pd.DataFrame(train_X_len), pd.DataFrame(train_X)], axis=1) \
        .to_csv(os.path.join(data_dir, 'train.csv'), header=False, index=False)

pd.concat([pd.DataFrame(test_X_len), pd.DataFrame(test_X)], axis=1) \
        .to_csv(os.path.join(data_dir, 'test.csv'), header=False, index=False)

>**Take-A-Look** Let's take a closer look at the concatenation above. The first 5 values oftrain_y are all 0's. See below. The first 5 review lengths are given in the one dimensional array below. The first 5 rows of the newly created numpy array are shown below by reading back the first 5 rows of the *.csv file. 

In [36]:
display (train_y[:5])
display (train_X_len[:5])

[0, 0, 0, 0, 0]

array([402, 106,  26,  77,  77])

In [37]:
# Just a look at the csv file: 
train=pd.read_csv(os.path.join(data_dir,'train.csv'))
oftype(train)
#display (train.head())


pandas DataFrame
(24999, 502)

The total number of missing values: 0 or 0.0%.         


0  402     1    12  3706  1165    30   130   776    44  ...  0.89  0.90  \
0  0  106   161     4     1  1005     2    56   842    32  ...     0     0   
1  0   26   254    12     3  3028   722    49    11  2377  ...     0     0   
2  0   77  1677  2699    75     1     1  2336     1  2226  ...     0     0   
3  0   77    47    46    46     3  2315  2287  1758   174  ...     0     0   
4  0  120  3030   123  4057    40    38  3030     3   263  ...     0     0   

   0.91  0.92  0.93  0.94  0.95  0.96  0.97  0.98  
0     0     0     0     0     0     0     0     0  
1     0     0     0     0     0     0     0     0  
2     0     0     0     0     0     0     0     0  
3     0     0     0     0     0     0     0     0  
4     0     0     0     0     0     0     0     0  

[5 rows x 502 columns]

### Uploading the training data


Next, we need to upload the training data to the SageMaker default S3 bucket so that we can provide access to it while training our model.

In [38]:
import sagemaker

session = sagemaker.Session()

bucket = session.default_bucket()
prefix = 'Toms-Project'

role = sagemaker.get_execution_role()

In [39]:
train_s3 = session.upload_data(os.path.join(data_dir,'train.csv'), 
                            bucket=bucket, 
                            key_prefix='{}'.format(prefix))
test_s3 = session.upload_data(os.path.join(data_dir,'test.csv'), 
                            bucket=bucket, 
                            key_prefix='{}'.format(prefix))

In [40]:
input_data = session.upload_data(path=data_dir, bucket=bucket, key_prefix=prefix)

In [41]:
len(input_data)
display (input_data)

's3://sagemaker-us-east-2-091975412881/Toms-Project'

In [42]:
# We are done with the local training datasets and they can be deleted: 

train_X = train_y = train_X_len =  None

**NOTE:** The cell above uploads the entire contents of our data directory. This includes the `word_dict.pkl` file. This is fortunate as we will need this later on when we create an endpoint that accepts an arbitrary review. For now, we will just take note of the fact that it resides in the data directory (and so also in the S3 training bucket) and that we will need to make sure it gets saved in the model directory.

## Step 4: Build and Train the PyTorch Model

In the XGBoost notebook we discussed what a model is in the SageMaker framework. In particular, a model comprises three objects

 - Model Artifacts,
 - Training Code, and
 - Inference Code,
 
each of which interact with one another. In the XGBoost example we used training and inference code that was provided by Amazon. Here we will still be using containers provided by Amazon with the added benefit of being able to include our own custom code.

We will start by implementing our own neural network in PyTorch along with a training script. For the purposes of this project we have provided the necessary model object in the `model.py` file, inside of the `train` folder. You can see the provided implementation by running the cell below.

In [43]:
#!pygmentize train/model.py

The important takeaway from the implementation provided is that there are three parameters that we may wish to tweak to improve the performance of our model. These are the embedding dimension, the hidden dimension and the size of the vocabulary. We will likely want to make these parameters configurable in the training script so that if we wish to modify them we do not need to modify the script itself. We will see how to do this later on. To start we will write some of the training code in the notebook so that we can more easily diagnose any issues that arise.

First we will load a small portion of the training data set to use as a sample. It would be very time consuming to try and train the model completely in the notebook as we do not have access to a gpu and the compute instance that we are using is not particularly powerful. However, we can work on a small bit of the data to get a feel for how our training script is behaving.

In [44]:
import torch
import torch.utils.data

# Read in only the first 250 rows
train_sample = pd.read_csv(os.path.join(data_dir, 'train.csv'), header=None, names=None, nrows=250)

# Turn the input pandas dataframe into tensors
train_sample_y = torch.from_numpy(train_sample[[0]].values).float().squeeze()
train_sample_X = torch.from_numpy(train_sample.drop([0], axis=1).values).long()

# Build the dataset
train_sample_ds = torch.utils.data.TensorDataset(train_sample_X, train_sample_y)
# Build the dataloader
train_sample_dl = torch.utils.data.DataLoader(train_sample_ds, batch_size=50)

In [45]:
# Read in only the first 250 rows
test_sample = pd.read_csv(os.path.join(data_dir, 'test.csv'), header=None, names=None, nrows=250)

# Turn the input pandas dataframe into tensors
test_sample_y = torch.from_numpy(test_sample[[0]].values).float().squeeze()
test_sample_X = torch.from_numpy(test_sample.drop([0], axis=1).values).long()

# Build the dataset
test_sample_ds = torch.utils.data.TensorDataset(train_sample_X, train_sample_y)
# Build the dataloader
test_sample_dl = torch.utils.data.DataLoader(train_sample_ds, batch_size=50)

In [46]:
display (train_sample.head())
#display (test_sample.head())

0    1     2     3     4     5     6     7     8     9    ...  492  493  \
0    0  402     1    12  3706  1165    30   130   776    44  ...    0    0   
1    0  106   161     4     1  1005     2    56   842    32  ...    0    0   
2    0   26   254    12     3  3028   722    49    11  2377  ...    0    0   
3    0   77  1677  2699    75     1     1  2336     1  2226  ...    0    0   
4    0   77    47    46    46     3  2315  2287  1758   174  ...    0    0   

   494  495  496  497  498  499  500  501  
0    0    0    0    0    0    0    0    0  
1    0    0    0    0    0    0    0    0  
2    0    0    0    0    0    0    0    0  
3    0    0    0    0    0    0    0    0  
4    0    0    0    0    0    0    0    0  

[5 rows x 502 columns]

### (TODO) Writing the training method

Next we need to write the training code itself. This should be very similar to training methods that you have written before to train PyTorch models. We will leave any difficult aspects such as model saving / loading and parameter loading until a little later.

In [47]:
def train(model, train_loader, epochs, optimizer, loss_fn, device):
    for epoch in range(1, epochs + 1):
        model.train()
        total_loss = 0
        for batch in train_loader:         
            batch_X, batch_y = batch
            
            batch_X = batch_X.to(device)
            batch_y = batch_y.to(device)
            
            # TODO: Complete this train method to train the model provided.
            optimizer.zero_grad()
            out = model(batch_X)
            loss = loss_fn(out, batch_y)
            loss.backward()
            optimizer.step()
            
            
            total_loss += loss.data.item()
        print("Epoch: {}, BCELoss: {}".format(epoch, round(total_loss / len(train_loader),2)))

Supposing we have the training method above, we will test that it is working by writing a bit of code in the notebook that executes our training method on the small sample training set that we loaded earlier. The reason for doing this in the notebook is so that we have an opportunity to fix any errors that arise early when they are easier to diagnose.

In [48]:
import torch.optim as optim
from train.model import LSTMClassifier

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = LSTMClassifier(32, 100, 5000).to(device)
optimizer = optim.Adam(model.parameters())
loss_fn = torch.nn.BCELoss()

train(model, train_sample_dl, 3, optimizer, loss_fn, device)

Epoch: 1, BCELoss: 0.69
Epoch: 2, BCELoss: 0.68
Epoch: 3, BCELoss: 0.68


In order to construct a PyTorch model using SageMaker we must provide SageMaker with a training script. We may optionally include a directory which will be copied to the container and from which our training code will be run. When the training container is executed it will check the uploaded directory (if there is one) for a `requirements.txt` file and install any required Python libraries, after which the training script will be run.

### (TODO) Training the model

When a PyTorch model is constructed in SageMaker, an entry point must be specified. This is the Python file which will be executed when the model is trained. Inside of the `train` directory is a file called `train.py` which has been provided and which contains most of the necessary code to train our model. The only thing that is missing is the implementation of the `train()` method which you wrote earlier in this notebook.

**TODO**: Copy the `train()` method written above and paste it into the `train/train.py` file where required.

The way that SageMaker passes hyperparameters to the training script is by way of arguments. These arguments can then be parsed and used in the training script. To see how this is done take a look at the provided `train/train.py` file.

In [49]:
from sagemaker.pytorch import PyTorch

estimator = PyTorch(entry_point="train.py",
                    source_dir="train",
                    role=role,
                    framework_version='1.5.0', #0.4.0',
                    py_version = 'py3',
                    train_instance_count=1,
                    train_instance_type='ml.m4.xlarge',
                    output_path='s3://{}/{}/output'.format(session.default_bucket(), prefix),
                    sagemaker_session=session,
                    hyperparameters={
                        'epochs': 10,
                        'hidden_dim': 200,
                    })

In [50]:
%%time
estimator.fit({'training': input_data})

'create_image_uri' will be deprecated in favor of 'ImageURIProvider' class in SageMaker Python SDK v2.
's3_input' class will be renamed to 'TrainingInput' in SageMaker Python SDK v2.
'create_image_uri' will be deprecated in favor of 'ImageURIProvider' class in SageMaker Python SDK v2.


2020-12-17 16:48:01 Starting - Starting the training job...
2020-12-17 16:48:02 Starting - Launching requested ML instances......
2020-12-17 16:49:05 Starting - Preparing the instances for training...
2020-12-17 16:49:52 Downloading - Downloading input data...
2020-12-17 16:50:27 Training - Downloading the training image...
2020-12-17 16:50:54 Training - Training image download completed. Training in progress..bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell
2020-12-17 16:50:55,708 sagemaker-containers INFO     Imported framework sagemaker_pytorch_container.training
2020-12-17 16:50:55,712 sagemaker-containers INFO     No GPUs detected (normal if no gpus installed)
2020-12-17 16:50:55,725 sagemaker_pytorch_container.training INFO     Block until all host DNS lookups succeed.
2020-12-17 16:50:58,765 sagemaker_pytorch_container.training INFO     Invoking user training script.
2020-12-17 16:50:59,124 sagemaker-containers INFO 

## Step 5: Testing the model

As mentioned at the top of this notebook, we will be testing this model by first deploying it and then sending the testing data to the deployed endpoint. We will do this so that we can make sure that the deployed model is working correctly.

## Step 6: Deploy the model for testing

Now that we have trained our model, we would like to test it to see how it performs. Currently our model takes input of the form `review_length, review[500]` where `review[500]` is a sequence of `500` integers which describe the words present in the review, encoded using `word_dict`. Fortunately for us, SageMaker provides built-in inference code for models with simple inputs such as this.

There is one thing that we need to provide, however, and that is a function which loads the saved model. This function must be called `model_fn()` and takes as its only parameter a path to the directory where the model artifacts are stored. This function must also be present in the python file which we specified as the entry point. In our case the model loading function has been provided and so no changes need to be made.

**NOTE**: When the built-in inference code is run it must import the `model_fn()` method from the `train.py` file. This is why the training code is wrapped in a main guard ( ie, `if __name__ == '__main__':` )

Since we don't need to change anything in the code that was uploaded during training, we can simply deploy the current model as-is.

**NOTE:** When deploying a model you are asking SageMaker to launch an compute instance that will wait for data to be sent to it. As a result, this compute instance will continue to run until *you* shut it down. This is important to know since the cost of a deployed endpoint depends on how long it has been running for.

In other words **If you are no longer using a deployed endpoint, shut it down!**

**TODO:** Deploy the trained model.

## Step 7 - Use the model for testing

Once deployed, we can read in the test data and send it off to our deployed model to get some results. Once we collect all of the results we can determine how accurate our model is.

>**Take-A-Look** Note that the training set was composed of the arrays train_y, train_X_len and train_X for a total of 502 columns. For testing, the dataset will contain test_X_len and test_X for a total of 501 columns:

In [51]:
oftype(test_X)
oftype(test_X_len)

numpy (25000, 500)
numpy (25000,)


In [52]:
test_X = pd.concat([pd.DataFrame(test_X_len), pd.DataFrame(test_X)], axis=1)

In [53]:
oftype(test_X)

pandas DataFrame
(25000, 501)

The total number of missing values: 0 or 0.0%.         


0     0    1    2     3     4    5     6    7     8    ...  490  491  492  \
0  234   815  311  657     2   698  521   338   34   185  ...    0    0    0   
1   80   427  165  705    31    82  681  4723    1    44  ...    0    0    0   
2   94   395    3   44    26     9   41    97  445  2964  ...    0    0    0   
3  209  2928   11  206  4111  1232   15   270  209     1  ...    0    0    0   
4   68    28   30   73    32   222  204    13  392    13  ...    0    0    0   

   493  494  495  496  497  498  499  
0    0    0    0    0    0    0    0  
1    0    0    0    0    0    0    0  
2    0    0    0    0    0    0    0  
3    0    0    0    0    0    0    0  
4    0    0    0    0    0    0    0  

[5 rows x 501 columns]

In [54]:
# We split the data into chunks and send each chunk seperately, accumulating the results.

#def predict(data, rows=512):
def predict(data, rows=1024):
    split_array = np.array_split(data, int(data.shape[0] / float(rows) + 1))
    predictions = np.array([])
    for array in split_array:
        predictions = np.append(predictions, predictor.predict(array))
    
    return predictions

In [55]:
predictor = estimator.deploy(initial_instance_count=1, instance_type='ml.m4.xlarge')

Parameter image will be renamed to image_uri in SageMaker Python SDK v2.
'create_image_uri' will be deprecated in favor of 'ImageURIProvider' class in SageMaker Python SDK v2.


---------------!

>**Take-A-Look** Note below we are using test_X.values. The .values converts the pandas dataframe into a numpy array, which is the object type required for the predict function. 

In [56]:
%%time
predictions = predict(test_X.values)
predictions = [round(num) for num in predictions]

CPU times: user 621 ms, sys: 35.9 ms, total: 657 ms
Wall time: 2min 40s


In [57]:
from sklearn.metrics import accuracy_score
accuracy_score(test_y, predictions)

0.851

**Question:** How does this model compare to the XGBoost model you created earlier? Why might these two models perform differently on this dataset? Which do *you* think is better for sentiment analysis?

**Answer:** From the accuracy scores, the models are similar. 

### (TODO) More testing

We now have a trained model which has been deployed and which we can send processed reviews to and which returns the predicted sentiment. However, ultimately we would like to be able to send our model an unprocessed review. That is, we would like to send the review itself as a string. For example, suppose we wish to send the following review to our model.

In [58]:
test_review = 'The simplest pleasures in life are the best, and this film is one of them. Combining a rather basic storyline of love and adventure this movie transcends the usual weekend fair with wit and unmitigated charm.'

The question we now need to answer is, how do we send this review to our model?

Recall in the first section of this notebook we did a bunch of data processing to the IMDb dataset. In particular, we did two specific things to the provided reviews.
 - Removed any html tags and stemmed the input
 - Encoded the review as a sequence of integers using `word_dict`
 
In order process the review we will need to repeat these two steps.

**TODO**: Using the `review_to_words` and `convert_and_pad` methods from section one, convert `test_review` into a numpy array `test_data` suitable to send to our model. Remember that our model expects input of the form `review_length, review[500]`.

In [59]:
# TODO: Convert test_review into a form usable by the model and save the results in test_data

words = review_to_words(test_review)
print (words)
#test_data = [np.array(convert_and_pad(word_dict, words)[0])]
#oftype(test_data)

#newtest_x, newtest_len = convert_and_pad(word_dict, words)
#oftype(newtest_x)
#oftype(newtest_len)

['simplest', 'pleasur', 'life', 'best', 'film', 'one', 'combin', 'rather', 'basic', 'storylin', 'love', 'adventur', 'movi', 'transcend', 'usual', 'weekend', 'fair', 'wit', 'unmitig', 'charm']


In [60]:
#test_data = [np.array(convert_and_pad(word_dict, words)[0])]
#test_X = pd.concat([pd.DataFrame(newtest_len), pd.DataFrame(newtest_x)], axis=1)
test_data = [np.array(convert_and_pad(word_dict, review_to_words(test_review))[0])]
oftype(test_data)
display (test_data)

List of length:  1


[array([   1, 1373,   50,   53,    3,    4,  878,  173,  392,  682,   29,
         724,    2, 4425,  275, 2075, 1059,  760,    1,  581,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0, 

Now that we have processed the review, we can send the resulting array to our model to predict the sentiment of the review.

In [61]:
predictor.predict(test_data)

array(0.52915746)

This value is low. Probably reflects the small sample size of the test_data. 

### Delete the endpoint

Of course, just like in the XGBoost notebook, once we've deployed an endpoint it continues to run until we tell it to shut down. Since we are done using our endpoint for now, we can delete it.

In [62]:
estimator.delete_endpoint()

estimator.delete_endpoint() will be deprecated in SageMaker Python SDK v2. Please use the delete_endpoint() function on your predictor instead.


In [63]:
oftype(test_data)

List of length:  1


## Step 6 (again) - Deploy the model for the web app

Now that we know that our model is working, it's time to create some custom inference code so that we can send the model a review which has not been processed and have it determine the sentiment of the review.

As we saw above, by default the estimator which we created, when deployed, will use the entry script and directory which we provided when creating the model. However, since we now wish to accept a string as input and our model expects a processed review, we need to write some custom inference code.

We will store the code that we write in the `serve` directory. Provided in this directory is the `model.py` file that we used to construct our model, a `utils.py` file which contains the `review_to_words` and `convert_and_pad` pre-processing functions which we used during the initial data processing, and `predict.py`, the file which will contain our custom inference code. Note also that `requirements.txt` is present which will tell SageMaker what Python libraries are required by our custom inference code.

When deploying a PyTorch model in SageMaker, you are expected to provide four functions which the SageMaker inference container will use.
 - `model_fn`: This function is the same function that we used in the training script and it tells SageMaker how to load our model.
 - `input_fn`: This function receives the raw serialized input that has been sent to the model's endpoint and its job is to de-serialize and make the input available for the inference code.
 - `output_fn`: This function takes the output of the inference code and its job is to serialize this output and return it to the caller of the model's endpoint.
 - `predict_fn`: The heart of the inference script, this is where the actual prediction is done and is the function which you will need to complete.

For the simple website that we are constructing during this project, the `input_fn` and `output_fn` methods are relatively straightforward. We only require being able to accept a string as input and we expect to return a single value as output. You might imagine though that in a more complex application the input or output may be image data or some other binary data which would require some effort to serialize.

### (TODO) Writing inference code

Before writing our custom inference code, we will begin by taking a look at the code which has been provided.

In [64]:
#!pygmentize serve/predict.py

As mentioned earlier, the `model_fn` method is the same as the one provided in the training code and the `input_fn` and `output_fn` methods are very simple and your task will be to complete the `predict_fn` method. Make sure that you save the completed file as `predict.py` in the `serve` directory.

**TODO**: Complete the `predict_fn()` method in the `serve/predict.py` file.

### Deploying the model

Now that the custom inference code has been written, we will create and deploy our model. To begin with, we need to construct a new PyTorchModel object which points to the model artifacts created during training and also points to the inference code that we wish to use. Then we can call the deploy method to launch the deployment container.

**NOTE**: The default behaviour for a deployed PyTorch model is to assume that any input passed to the predictor is a `numpy` array. In our case we want to send a string so we need to construct a simple wrapper around the `RealTimePredictor` class to accomodate simple strings. In a more complicated situation you may want to provide a serialization object, for example if you wanted to sent image data.

In [65]:
%%time
from sagemaker.predictor import RealTimePredictor
import sagemaker.predictor
from sagemaker.pytorch import PyTorchModel

class StringPredictor(RealTimePredictor):
    def __init__(self, endpoint_name, sagemaker_session):
        super(StringPredictor, self).__init__(endpoint_name, sagemaker_session, content_type='text/plain')
        #super(StringPredictor, self).__init__(endpoint_name, sagemaker_session, content_type='multipart/mixed')
        #super(StringPredictor, self).__init__(endpoint_name, sagemaker_session, content_type='text/csv')

model = PyTorchModel(model_data=estimator.model_data,
                     role = role,
                     framework_version='1.5.0', #'0.4.0',
                     entry_point='predict.py',
                     py_version ='py3',
                     source_dir='serve',
                     predictor_cls=StringPredictor)
predictor = model.deploy(initial_instance_count=1, instance_type='ml.m4.xlarge')

Parameter image will be renamed to image_uri in SageMaker Python SDK v2.
'create_image_uri' will be deprecated in favor of 'ImageURIProvider' class in SageMaker Python SDK v2.


-----------------!CPU times: user 649 ms, sys: 20 ms, total: 669 ms
Wall time: 8min 33s


### Testing the model

Now that we have deployed our model with the custom inference code, we should test to see if everything is working. Here we test our model by loading the first `250` positive and negative reviews and send them to the endpoint, then collect the results. The reason for only sending some of the data is that the amount of time it takes for our model to process the input and then perform inference is quite long and so testing the entire data set would be prohibitive.

In [66]:
import glob

def test_reviews(data_dir='../data/aclImdb', stop=250):
    
    results = []
    ground = []
    
    # We make sure to test both positive and negative reviews    
    for sentiment in ['pos', 'neg']:
        
        path = os.path.join(data_dir, 'test', sentiment, '*.txt')
        #print ("path: ",path)
        files = glob.glob(path)
        
        files_read = 0
        
        #print('Starting ', sentiment, ' files')
        
        # Iterate through the files and send them to the predictor
        for f in files:
            #print (f)
            with open(f) as review:
                # First, we store the ground truth (was the review positive or negative)
                if sentiment == 'pos':
                    ground.append(1)
                else:
                    ground.append(0)
                #print ('appending ground completed')
                # Read in the review and convert to 'utf-8' for transmission via HTTP
                review_input = review.read().encode('utf-8')
                #print ('\n Review Input completed')
                # Send the review to the predictor and store the results
                results.append(int(predictor.predict(review_input)))
                #print ('results appended')
            # Sending reviews to our endpoint one at a time takes a while so we
            # only send a small number of reviews
            files_read += 1
            if files_read == stop:
                break
            
    return ground, results

In [67]:
ground, results = test_reviews()

In [68]:
from sklearn.metrics import accuracy_score
accuracy_score(ground, results)

0.858

As an additional test, we can try sending the `test_review` that we looked at earlier.

In [69]:
predictor.predict(test_review)

b'1'

Now that we know our endpoint is working as expected, we can set up the web page that will interact with it. If you don't have time to finish the project now, make sure to skip down to the end of this notebook and shut down your endpoint. You can deploy it again when you come back.

## Step 7 (again): Use the model for the web app

> **TODO:** This entire section and the next contain tasks for you to complete, mostly using the AWS console.

So far we have been accessing our model endpoint by constructing a predictor object which uses the endpoint and then just using the predictor object to perform inference. What if we wanted to create a web app which accessed our model? The way things are set up currently makes that not possible since in order to access a SageMaker endpoint the app would first have to authenticate with AWS using an IAM role which included access to SageMaker endpoints. However, there is an easier way! We just need to use some additional AWS services.

<img src="Web App Diagram.svg">

The diagram above gives an overview of how the various services will work together. On the far right is the model which we trained above and which is deployed using SageMaker. On the far left is our web app that collects a user's movie review, sends it off and expects a positive or negative sentiment in return.

In the middle is where some of the magic happens. We will construct a Lambda function, which you can think of as a straightforward Python function that can be executed whenever a specified event occurs. We will give this function permission to send and recieve data from a SageMaker endpoint.

Lastly, the method we will use to execute the Lambda function is a new endpoint that we will create using API Gateway. This endpoint will be a url that listens for data to be sent to it. Once it gets some data it will pass that data on to the Lambda function and then return whatever the Lambda function returns. Essentially it will act as an interface that lets our web app communicate with the Lambda function.

### Setting up a Lambda function

The first thing we are going to do is set up a Lambda function. This Lambda function will be executed whenever our public API has data sent to it. When it is executed it will receive the data, perform any sort of processing that is required, send the data (the review) to the SageMaker endpoint we've created and then return the result.

#### Part A: Create an IAM Role for the Lambda function

Since we want the Lambda function to call a SageMaker endpoint, we need to make sure that it has permission to do so. To do this, we will construct a role that we can later give the Lambda function.

Using the AWS Console, navigate to the **IAM** page and click on **Roles**. Then, click on **Create role**. Make sure that the **AWS service** is the type of trusted entity selected and choose **Lambda** as the service that will use this role, then click **Next: Permissions**.

In the search box type `sagemaker` and select the check box next to the **AmazonSageMakerFullAccess** policy. Then, click on **Next: Review**.

Lastly, give this role a name. Make sure you use a name that you will remember later on, for example `LambdaSageMakerRole`. Then, click on **Create role**.

#### Part B: Create a Lambda function

Now it is time to actually create the Lambda function.

Using the AWS Console, navigate to the AWS Lambda page and click on **Create a function**. When you get to the next page, make sure that **Author from scratch** is selected. Now, name your Lambda function, using a name that you will remember later on, for example `sentiment_analysis_func`. Make sure that the **Python 3.6** runtime is selected and then choose the role that you created in the previous part. Then, click on **Create Function**.

On the next page you will see some information about the Lambda function you've just created. If you scroll down you should see an editor in which you can write the code that will be executed when your Lambda function is triggered. In our example, we will use the code below. 

```python
# We need to use the low-level library to interact with SageMaker since the SageMaker API
# is not available natively through Lambda.
import boto3

def lambda_handler(event, context):

    # The SageMaker runtime is what allows us to invoke the endpoint that we've created.
    runtime = boto3.Session().client('sagemaker-runtime')

    # Now we use the SageMaker runtime to invoke our endpoint, sending the review we were given
    response = runtime.invoke_endpoint(EndpointName = '**ENDPOINT NAME HERE**',    # The name of the endpoint we created
                                       ContentType = 'text/plain',                 # The data format that is expected
                                       Body = event['body'])                       # The actual review

    # The response is an HTTP response whose body contains the result of our inference
    result = response['Body'].read().decode('utf-8')

    return {
        'statusCode' : 200,
        'headers' : { 'Content-Type' : 'text/plain', 'Access-Control-Allow-Origin' : '*' },
        'body' : result
    }
```

Once you have copy and pasted the code above into the Lambda code editor, replace the `**ENDPOINT NAME HERE**` portion with the name of the endpoint that we deployed earlier. You can determine the name of the endpoint using the code cell below.

In [70]:
predictor.endpoint

'pytorch-inference-2020-12-17-17-45-33-781'

In [71]:
#print (str(vocabulary)) 

Once you have added the endpoint name to the Lambda function, click on **Save**. Your Lambda function is now up and running. Next we need to create a way for our web app to execute the Lambda function.

### Setting up API Gateway

Now that our Lambda function is set up, it is time to create a new API using API Gateway that will trigger the Lambda function we have just created.

Using AWS Console, navigate to **Amazon API Gateway** and then click on **Get started**.

On the next page, make sure that **New API** is selected and give the new api a name, for example, `sentiment_analysis_api`. Then, click on **Create API**.

Now we have created an API, however it doesn't currently do anything. What we want it to do is to trigger the Lambda function that we created earlier.

Select the **Actions** dropdown menu and click **Create Method**. A new blank method will be created, select its dropdown menu and select **POST**, then click on the check mark beside it.

For the integration point, make sure that **Lambda Function** is selected and click on the **Use Lambda Proxy integration**. This option makes sure that the data that is sent to the API is then sent directly to the Lambda function with no processing. It also means that the return value must be a proper response object as it will also not be processed by API Gateway.

Type the name of the Lambda function you created earlier into the **Lambda Function** text entry box and then click on **Save**. Click on **OK** in the pop-up box that then appears, giving permission to API Gateway to invoke the Lambda function you created.

The last step in creating the API Gateway is to select the **Actions** dropdown and click on **Deploy API**. You will need to create a new Deployment stage and name it anything you like, for example `prod`.

You have now successfully set up a public API to access your SageMaker model. Make sure to copy or write down the URL provided to invoke your newly created public API as this will be needed in the next step. This URL can be found at the top of the page, highlighted in blue next to the text **Invoke URL**.

## Step 4: Deploying our web app

Now that we have a publicly available API, we can start using it in a web app. For our purposes, we have provided a simple static html file which can make use of the public api you created earlier.

In the `website` folder there should be a file called `index.html`. Download the file to your computer and open that file up in a text editor of your choice. There should be a line which contains **\*\*REPLACE WITH PUBLIC API URL\*\***. Replace this string with the url that you wrote down in the last step and then save the file.

Now, if you open `index.html` on your local computer, your browser will behave as a local web server and you can use the provided site to interact with your SageMaker model.

If you'd like to go further, you can host this html file anywhere you'd like, for example using github or hosting a static site on Amazon's S3. Once you have done this you can share the link with anyone you'd like and have them play with it too!

> **Important Note** In order for the web app to communicate with the SageMaker endpoint, the endpoint has to actually be deployed and running. This means that you are paying for it. Make sure that the endpoint is running when you want to use the web app but that you shut it down when you don't need it, otherwise you will end up with a surprisingly large AWS bill.

**TODO:** Make sure that you include the edited `index.html` file in your project submission.

Now that your web app is working, trying playing around with it and see how well it works.

**Question**: Give an example of a review that you entered into your web app. What was the predicted sentiment of your example review?

**Answer:**

### Delete the endpoint

Remember to always shut down your endpoint if you are no longer using it. You are charged for the length of time that the endpoint is running so if you forget and leave it on you could end up with an unexpectedly large bill.

In [72]:
predictor.delete_endpoint()